antelope_contracts/contracts/erc20/src/erc20.cpp

```diff
@@ -8,92 +8,269 @@
+    auto pack_uint256 = [&](bytes &ds, const intx::uint256 &val) {
+        uint8_t val_[32] = {};
+        intx::be::store(val_, val);
+        ds.insert(ds.end(), val_, val_ + sizeof(val_));
     };
+    auto pack_uint32 = [&](bytes &ds, uint32_t val) {
+        uint8_t val_[32] = {};
+        val_[28] = (uint8_t)(val >> 24);
+        val_[29] = (uint8_t)(val >> 16);
+        val_[30] = (uint8_t)(val >> 8);
+        val_[31] = (uint8_t)val;
+        ds.insert(ds.end(), val_, val_ + sizeof(val_));
+    };
+    auto pack_string = [&](bytes &ds, const auto &str) {
+        pack_uint32(ds, (uint32_t)str.size());
+        for (size_t i = 0; i < (str.size() + 31) / 32 * 32; i += 32) {
+            uint8_t name_[32] = {};
+            memcpy(name_, str.data() + i, i + 32 > str.size() ? str.size() - i : 32);
+            ds.insert(ds.end(), name_, name_ + sizeof(name_));
+        }
+    };
+
+    pack_uint32(constructor_data, (uint8_t)erc20_precision); // offset 0
+    pack_uint256(constructor_data, egress_fee_evm);          // offset 32
+    pack_uint32(constructor_data, 160);                      // offset 64
+    pack_uint32(constructor_data, 224);                      // offset 96
+    pack_uint32(constructor_data, 288);                      // offset 128
+    pack_string(constructor_data, evm_token_name);           // offset 160
+    pack_string(constructor_data, evm_token_symbol);         // offset 224
+    pack_string(constructor_data, eos_contract_name.to_string()); // offset 288
 
-    new_config.erc20_addr.resize(kAddressLength);
-    memcpy(new_config.erc20_addr.data(), deploy_addr.bytes, kAddressLength);
+    pack_uint32(call_data, 64);                  // offset 32
+    pack_string(call_data, constructor_data);    // offset 64
 
-    _config.set(new_config, get_self());
+    bytes to = {};
+    bytes value_zero; 
+    value_zero.resize(32, 0);
+
+     // required account opened in evm_runtime
+    call_action call_act(evm_account, {{get_self(), "active"_n}});
+    call_act.send(get_self(), to, value_zero, call_data, evm_init_gaslimit);
+
+    evmc::address proxy_contract_addr = silkworm::create_address(reserved_addr, my_next_nonce()); 
+
+    token_table.emplace(_self, [&](auto &v) {
+        v.eos_contract_name = eos_contract_name;
+        v.address.resize(kAddressLength, 0);
+        memcpy(&(v.address[0]), proxy_contract_addr.bytes, kAddressLength);
+        v.min_deposit = min_deposit;
+        v.deposit_fee = deposit_fee;
+        v.erc20_precision = erc20_precision;
+    });
 }
 
-void erc20::onbridgemsg(name receiver, const bytes& sender, const time_point& timestamp, const bytes& value, const bytes& data) {
-    require_auth(evm_account);
+void erc20::onbridgemsg(const bridge_message_t &message) {
+
+    check(get_sender() == evm_account, "invalid sender of onbridgemsg");
+
+    const bridge_message_v0 &msg = std::get<bridge_message_v0>(message);
+    check(msg.receiver == _self, "invalid message receiver");
```

Comment:
This should not be `_self` but rather the variable holding the `eosio.erc2o` name.